pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,473 @@
+	if err != nil {
+		return "", err
+	}
+
+	for _, tenant := range tenants.Items {
+		if tenant.Name == tenantName {
+			projects, err := Esxclient.Tenants.GetProjects(tenant.ID, nil)
+			if err != nil {
+				return "", err
+			}
+
+			for _, project := range projects.Items {
+				if project.Name == projName {
+					return project.ID, nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("Photon Cloud Provider: no matching tenant/project name is found with %s/%s", tenantName, projName)
+}
+
+func newPCCloud(cfg PCConfig) (*PCCloud, error) {
+	if len(cfg.Global.CloudTarget) == 0 {
+		return nil, errors.New("Invalid Photon Controller endpoint")
+	}
+
+	//TODO: add handling of certification enabled situation
+	options := &photon.ClientOptions{
+		IgnoreCertificate: cfg.Global.IgnoreCertificate,
+	}
+
+	Esxclient = photon.NewClient(cfg.Global.CloudTarget, options, logger)
+	status, err := Esxclient.Status.Get()
+	if err != nil {
+		return nil, errors.New("Photon Controller new client creation failed")
+	}
+	glog.V(2).Info("Photon Cloud Provider: Status of the new photon controller client: %v", status)
+
+	// Get Photon Controller project ID for future use
+	projID, err := getProjIDbyName(cfg.Global.Tenant, cfg.Global.Project)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get local hostname for localInstanceID
+	cmd := exec.Command("bash", "-c", `echo $HOSTNAME`)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err = cmd.Run()
+	if err != nil {
+		return nil, err
+	}
+	if out.Len() == 0 {
+		return nil, fmt.Errorf("Photon Cloud Provider: unable to retrieve hostname for Instance ID")
+	}
+	hostname := strings.TrimRight(out.String(), "\n")
+	vmID, err := getVMIDbyNodename(projID, hostname)
+	if err != nil {
+		return nil, err
+	}
+
+	pc := PCCloud{
+		cfg:              &cfg,
+		localInstanceID:  vmID,
+		localHostname:    hostname,
+		localK8sHostname: "",
+		projID:           projID,
+	}
+
+	overrideIP = cfg.Global.OverrideIP
+
+	return &pc, nil
+}
+
+// Instances returns an implementation of Instances for Photon Controller.
+func (pc *PCCloud) Instances() (cloudprovider.Instances, bool) {
+	return pc, true
+}
+
+// List is an implementation of Instances.List.
+func (pc *PCCloud) List(filter string) ([]k8stypes.NodeName, error) {
+	return nil, nil
+}
+
+// NodeAddresses is an implementation of Instances.NodeAddresses.
+func (pc *PCCloud) NodeAddresses(nodeName k8stypes.NodeName) ([]api.NodeAddress, error) {
+	addrs := []api.NodeAddress{}
+	return addrs, nil
+}
+
+func (pc *PCCloud) AddSSHKeyToAllInstances(user string, keyData []byte) error {
+	return errors.New("unimplemented")
+}
+
+func (pc *PCCloud) CurrentNodeName(hostname string) (k8stypes.NodeName, error) {
+	pc.localK8sHostname = hostname
+	return k8stypes.NodeName(hostname), nil
+}
+
+// ExternalID returns the cloud provider ID of the specified instance (deprecated).
+func (pc *PCCloud) ExternalID(nodeName k8stypes.NodeName) (string, error) {
+	return "ExtID", nil
```

Comment:
This is also a deprecated function. But seems it's used by one function to check if the node is existing or not. It should be replaced by InstanceID function. For now I will just repeat the InstanceID function implementation for it.